src/compiler/compile/render_dom/index.ts

```diff
@@ -246,7 +254,12 @@ export default function dom(
 	}
 
 	const args = [x`$$self`];
-	if (props.length > 0 || component.has_reactive_assignments || component.slots.size > 0) {
+	const has_invalidate = props.length > 0 ||
+		component.has_reactive_assignments ||
+		component.slots.size > 0 ||
+		capture_state ||
```

Comment:
Is the intent here (and then again below) to check whether there's a non-noop capture_state/inject_state? Right now, won't this will be truthy whenever we're compiling in dev mode?

Reply:
Hm when I made a change to this effect locally, what ended up happening was that in the `debug-no-dependencies` test, no `instance` function was generated at all, and so `$$self.$capture_state`/`$$self.$inject_state` weren't set, even to `noop`. So perhaps what's here now is correct.

Reply:
An alternative would be to make `$capture_state` and `$inject_state` actually exist as instance methods on `SvelteComponentDev` (as noops), and have them be overridden in `instance` when necessary - which I might actually like better. Thoughts?

Reply:
Hey, thanks :)

So yeah, it was always true in dev mode (before your change) but I just wanted to write out the real reason, that it was needed for `$capture_state` and `$inject_state`.

I agree that having them on `SvelteComponentDev` is far better. It also explicits these methods' contract by making it clear that they are always present for all components in dev mode. It's great.